spec/abstract/Sporadic_A.thy

```diff
@@ -0,0 +1,121 @@
+    sc \<leftarrow> get_sched_context sc_ptr;
+    return (sc_period sc = 0)
+  od"
+
+definition
+  refill_pop_head :: "obj_ref \<Rightarrow> (refill, 'z::state_ext) s_monad"
+where
+  "refill_pop_head sc_ptr \<equiv> do
+     refills \<leftarrow> get_refills sc_ptr;
+     update_sched_context sc_ptr (sc_refills_update tl);
+     return (hd refills)
+   od"
+
+definition
+  refill_head_overlapping :: "obj_ref \<Rightarrow> (bool, 'z::state_ext) r_monad"
+where
+  "refill_head_overlapping sc_ptr \<equiv> do {
+    sc \<leftarrow> read_sched_context sc_ptr;
+    oreturn (length (sc_refills sc) > 1
+             \<and> r_time (hd (tl (sc_refills sc))) \<le> r_time (refill_hd sc) + r_amount (refill_hd sc))
+  }"
+
+definition
+  update_refill_hd :: "obj_ref \<Rightarrow> (refill \<Rightarrow> refill) \<Rightarrow> (unit, 'z::state_ext) s_monad"
+where
+  "update_refill_hd sc_ptr f = update_sched_context sc_ptr (sc_refills_update (\<lambda>refills. f (hd refills) # (tl refills)))"
+
+definition
+  "can_merge_refill r1 r2 \<equiv> r_time r2 \<le> r_time r1 + r_amount r1"
+
+definition
+  merge_refill :: "refill \<Rightarrow> refill \<Rightarrow> refill"
+where
+  "merge_refill r1 r2 = \<lparr> r_time = r_time r1, r_amount = r_amount r2 + r_amount r1 \<rparr>"
+
+definition
+  merge_refills :: "obj_ref \<Rightarrow> (unit, 'z::state_ext) s_monad"
+where
+  "merge_refills sc_ptr \<equiv> do
+     head \<leftarrow> refill_pop_head sc_ptr;
+     update_refill_hd sc_ptr (merge_refill head)
+   od"
+
+definition
+  refill_head_overlapping_loop :: "obj_ref \<Rightarrow> (unit, 'z::state_ext) s_monad"
+where
+  "refill_head_overlapping_loop sc_ptr
+     \<equiv> whileLoop (\<lambda>_ s. the ((refill_head_overlapping sc_ptr) s)) (\<lambda>_. merge_refills sc_ptr) ()"
+
+definition
+  refill_unblock_check :: "obj_ref \<Rightarrow> (unit, 'z::state_ext) s_monad"
+where
+  "refill_unblock_check sc_ptr \<equiv> do
+    robin \<leftarrow> is_round_robin sc_ptr;
+    ready \<leftarrow> get_sc_refill_ready sc_ptr;
+    when (ready \<and> \<not>robin) $ do
+      modify (\<lambda>s. s\<lparr> reprogram_timer := True \<rparr>);
+      ct \<leftarrow> gets cur_time;
+      update_refill_hd sc_ptr (r_time_update (\<lambda>_. ct + kernelWCET_ticks));
+      refill_head_overlapping_loop sc_ptr
+    od
+  od"
+
+(* below currently gives if-test (when) for assert *)
+definition if_cond_refill_unblock_check where
+  "if_cond_refill_unblock_check sc_opt active asst \<equiv>
+    maybeM (\<lambda>scp. do
+      sc \<leftarrow> get_sched_context scp;
+      cur_sc_ptr \<leftarrow> gets cur_sc;
+      guard \<leftarrow> return (case active of
+                         None \<Rightarrow> (\<not> sc_sporadic sc)
+                       | Some True \<Rightarrow> sc_sporadic sc \<and> sc_active sc
+                       | Some False \<Rightarrow> sc_sporadic sc \<and> sc_active sc);
+      when (guard \<and> (if asst = Some False then scp \<noteq> cur_sc_ptr else True)) $
```

Comment:
Would it be better to write this `if _ then _ else True` as `_ --> _`, and the one on the next line?

Reply:
what kind of benefits do you have in mind?

Reply:
This is a wrapper function for `refill_unblock_check`, just adding these guards to the main function. So, when these guard conditions are irrelevant to what you are proving, you can mostly crunch things away. And, when the guards do matter for the proof, you are most likely to be wanting to split things apart. So, having if-then-else in this case might be in a sense effort saving, or at least the benefit mentioned above is not as much as what you might think. But I can certainly change it to an implication, too, maybe for the sake of consistency.

Reply:
And also, in actual call sites, the wrapper function is always called with concrete parameters, so these `if`s simplify away and there is no problem. It is only when we want to prove something generic about it that matters, and that is what my comment above is about.

Reply:
Actually, for this particular case, aggressive splitting won't be an issue because the simplifier will just change “if `P` then `X` else `True`” to `P --> X`. We can certainly use this result of simplification in the spec from the start, but I seem to have preference in spelling things out mainly due to the convoluted nature of the spec and also the possibility of the future code change, though it is not a very strong preference/argument.

Reply:
`the simplifier will just change “if P then X else True” to P --> X` is not correct -- the simplifier will usually pull out the implication to the top of the goal, which is not what you usually want, esp if your goal is a Hoare triple or similar.

Reply:
(even without that, I find `A --> B` a lot clearer and easier to read than `if A then B else True`)

Reply:
Hmm. Well, I had a look at the wp rules I proved, and apparently, what actually happens is that it doesn’t rewrite anything but it does not split this `if` either. Maybe because it is inside the `when` or with the conjunction, or maybe because of what is preceding I am not sure, but, for the case here, these `if`s are not being pulled out as implications. Anyway, happy to rewrite them as implications.

But I am a bit puzzled, too. This is in ASpec and I had the impression that we aren’t really trying to avoid if-then-else, generally. Maybe we should be doing that anyway.